irc/client.py

```diff
@@ -559,27 +559,49 @@ def as_nick(self, name):
             self.nick(orig)
 
     def process_data(self):
-        "read and process input from self.socket"
+        """"
+        Read and process input from self.socket
+        """
 
-        try:
-            reader = getattr(self.socket, 'read', self.socket.recv)
-            new_data = reader(2 ** 14)
-        except socket.error:
-            # The server hung up.
-            self.disconnect("Connection reset by peer")
-            return
-        if not new_data:
-            # Read nothing: connection must be down.
-            self.disconnect("Connection reset by peer")
-            return
+        # Read bytes until we got a line or a full buffer:
+        while len(self.buffer) < 600 and self.buffer.find(b"\n") < 0:
+            try:
+                reader = getattr(self.socket, 'read', self.socket.recv)
+                new_data = reader(1)
+            except socket.error:
+                # The server hung up.
+                self.disconnect("Connection reset by peer")
+                return
+            if not new_data:
+                # Read nothing: connection must be down.
+                self.disconnect("Connection reset by peer")
+                return
 
-        self.buffer.feed(new_data)
+            self.buffer.feed(new_data)
+
+        # If we got no line, the server is violating the protocol:
+        if self.buffer.find(b"\n") < 0:
+            self.buffer = b""
+            return
 
-        # process each non-empty line after logging all lines
-        for line in self.buffer:
-            log.debug("FROM SERVER: %s", line)
-            if not line: continue
-            self._process_line(line)
+        # Process each non-empty line:
+        next_break = self.buffer.find(b"\n")
+        while next_break >= 0:
+            next_line = self.buffer[:next_break]
+            if next_line.endswith("\r"):
+                next_line = next_line[:-1]
+            self.buffer = self.buffer[next_break:]
+            log.debug("FROM SERVER: %s", next_line)
+            decoded_line = None
+            try:
```

Comment:
This decoding technique is hard-coded, violating the explicit hooks advertised for customizing the encoding handling, and making it impossible to decode other encodings.